### DISCUSSION TITLE
Primal Bound and Integral

### DISCUSSION THREAD

**Discussion by AntoinePrv** [ds4dm/ecole#109] [Reward Metrics]
We'd like to implement a primal integral reward function.
What's a good definition/ implementation for it?

**Comment by AntoinePrv** [comment]
An idea to deal with Primal Integral not being defined as long as no solution are found could be to split the reward functions in two.
 - Primal integral would return `+inf` while there are no solutions
 - A second reward function would count the number of available solutions
 That way users can defined their own metric by combining the two, for instance:
 
 ```python
has_no_solution = ecole.reward.NSolutions() == 0
penalty = 1000
my_primal = min(penalty * has_no_solution + ecole.reward.PrimalIntegral())
```

@gasse @dchetelat what do you think?

**Reply by gasse** [reply]
I like the idea.

**Comment by AntoinePrv** [comment]
Sharing the issues about using `inf` that we discussed in the Ecole meeting.

If we have the SCIP primal bounds:
```
inf, inf, inf, 3.4, 3.3, 3.3, 2.1, 1.1
```
And in Ecole we want to provide rewards functions as differential improvements, then we would return
```
inf, nan, nan, -inf, -0.1, -0, -1.2, -1.0
```
Which completely erases the `3.4` offset, that is, using `PrimalBound.cumsum()` (or `DualBound().inf_to_num(0.).cumsum()` _etc_.), we can never recover the final primal bound.

Should we
 - a) Provide different reward function (getting the SCIP metric without difference is just a lambda)?
 - b) Or do weird arithmetics of `inf - inf` sort of things?
  
 I think I like a) better. It's the solution of least effort and not using any conventions. Something like
 ```python
 reward_functions = ecole.reward.lambda(lambda model: model.as_pyscipopt().getPrimalBound())
 ```
 
 cc @dchetelat @gasse @jdumouchelle @beljadid

**Reply by beljadid** [reply]
Hi,
If we have the SCIP primal bounds:
```
inf, inf, inf, 3.4, 3.3, 3.3, 2.1, 1.1, -inf (unbounded)
```

and if Ecole provide rewards functions as differential improvements, then we will have
```
inf, nan, nan, -inf, -0.1, -0, -1.2, -1.0,-inf
```

With differential improvements, i suggest (2 stages): 
inf, inf, inf,  (if the value is inf the reward function return inf)
```
3.4 (first value <> inf is returned), -0.1, -0, -1.2, -1.0,-inf
```

cc @dchetelat @gasse @jdumouchelle @beljadid @AntoinePrv

**Reply by gasse** [reply]
Maybe we should list the things we want for this reward function:
1. should `primalBoundDiff.cumsum().diff() == primalBoundDiff` ?
2. should `primalBoundDiff.cumsum() == SCIPgetPrimalBound()` ? We have this behavior with other reward functions, like `NNodes` and `LPIterations`
3. should we return NaN or Inf to the users, so that it is implicitly forbidden to use those reward functions as-is in their RL code ?
4. should we return reward values whose RL optimal policy is also optimal w.r.t. the Primal Bound ?
5. should we return reward values which are practical for RL ?

Note that @beljadid 's solution: returning 0 until a primal solution is found, then the first primal bound, and then the difference to the next primal bound, ensures 4., and 1. after a primal solution is found.

Primal bound:
`inf, inf, 3.4, 3.3, 3.3, 2.1, 1.1, -inf (unbounded)`

Reward:
`0.0, 0.0, 3.4, -0.1, 0.0, -1.2, -1.0, -inf`

Reward.cumsum():
`0.0, 0.0, 3.4, 3.3, 3.3, 2.1, 1.1, -inf`

Reward.cumsum().diff():
`0.0, 0.0, 3.4, -0.1, 0.0, -1.2, -1.0, -inf`

**Reply by AntoinePrv** [reply]
What is the first primal bound found is, in fact, `0.0`. Users won't be able to tell the difference with no primal bound.

**Reply by beljadid** [reply]
i think inf and -inf are mostely used to describe a state and not (rarely) used as values.
inf  :  not yet feasible
-inf : unbounded

the sum (integral) can be : spent_time_before_First_value * BigM * FirstValue +....
